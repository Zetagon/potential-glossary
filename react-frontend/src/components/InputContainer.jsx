/*jslint esversion:6*/
import React, { Component } from 'react';
import logo from '../logo.svg';
import '../App.css';
import '../../public/css/client-glos.css';

//eslint-disable-next-line
class InputBox extends Component {
  constructor(props){
    super(props);
    this.focus = this.focus.bind(this);
  }


  focus() {
    this.textInput.focus();
  }


  render(){
    let divStyle = {
      background: this.props.background_color
    };
    return (
      <div
        className="input-box"
        style={divStyle}>
        <input
          className="the-input"
          tabIndex="1"
          ref={ this.props.inputRef}
        />
      </div>
    );
  }
}

const CorrectColorMap = {
  Incorrect:'red',
  Correct:'green',
  NotSet:'white'
}

//eslint-disable-next-line
class InputContainer extends Component {

  constructor(props) {
    super(props);

    this.state = {
      inputBoxStatus: new Array(this.props.numberOfBoxes).fill('NotSet')
    }
    this.inputBoxRef = [];/* An array of references to inputbox dom-elements */ 
    this.highLightInputBoxes = this.highLightInputBoxes.bind(this);
  }

  focusFirstIncorrect() {
    for( let i = 0; i < this.inputBoxRef.length; i++) {
      if ( this.state.inputBoxStatus[i] != 'Correct') {
        this.inputBoxRef[i].focus();
        return
      }
    }
  }

  renderSubmitButton(){
    return(
      <div
        className="submit-button"
        onClick={() => this.submitButtonOnClick()
        }> Submit! </div>
    );
  }

  submitButtonOnClick(){
    this.highLightInputBoxes([
      'Correct',
      'Correct',
      'Correct',
      'NotSet',
      'Incorrect'
    ], () => this.focusFirstIncorrect());
  }

  renderInputboxes(){
    let boxes = new Array(this.props.numberOfBoxes).fill(null);
    boxes = boxes.map((x, index) => {
      return <InputBox
               key={ index }
               /* give a callback so that we can get a reference to the inputbox DOM element */
               inputRef={ el => this.inputBoxRef.push(el)} 
               background_color={
                 CorrectColorMap[this.state.inputBoxStatus[index]]
               } />;
    });
    return boxes;
  }

  highLightInputBoxes(ary, callback) {
    if (callback){
      this.setState({
        inputBoxStatus:ary
      }, callback);
    } else {
      this.setState({
        inputBoxStatus:ary
      })
    }
  }

  render(){
    let boxes = this.renderInputboxes();
    return(
      <div className="input-container">
        <div className="input-flex-container">
          { boxes }
        </div>
        {this.renderSubmitButton()}
      </div>);
  }
}

export {InputContainer, InputBox }
