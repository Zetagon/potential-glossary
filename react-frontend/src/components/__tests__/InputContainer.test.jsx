import React from 'react';
import {InputContainer, InputBox } from '../InputContainer.jsx'
import renderer from 'react-test-renderer';

test('number of boxes equal to synonyms', ()=> {
  let answer = {
    synonyms: [
      {
        alternatives: [
          {text: "hej"},
          {text: "hejsan"}
        ]
      },
      {
        alternatives: [
          {text: "god kväll"},
          {text: "god afton"}
        ]
      }
    ]
  } 

  const component = renderer.create(
    <InputContainer numberOfBoxes={ answer.synonyms.length }/>
  )

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  answer = {
    synonyms: [
      {
        alternatives: [
          {text: "hallao"},
          {text: "nej"}
        ]
      },
      {
        alternatives: [
          {text: "hallao"},
          {text: "hejsan"}
        ]
      },
      {
        alternatives: [
          {text: "hej"},
          {text: "hejsan"}
        ]
      },
      {
        alternatives: [
          {text: "god kväll"},
          {text: "god afton"}
        ]
      }
    ]
  } 

  const component1 = renderer.create(
    <InputContainer numberOfBoxes={ answer.synonyms.length }/>
  )

  let tree = component1.toJSON();
  expect(tree).toMatchSnapshot();
});
