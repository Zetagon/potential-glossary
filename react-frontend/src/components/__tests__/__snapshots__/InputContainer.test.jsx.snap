exports[`test number of boxes equal to synonyms 1`] = `
<div
  className="input-container">
  <div
    className="input-flex-container">
    <div
      className="input-box"
      style={
        Object {
          "background": "white",
        }
      }>
      <input
        className="the-input"
        tabIndex="1" />
    </div>
    <div
      className="input-box"
      style={
        Object {
          "background": "white",
        }
      }>
      <input
        className="the-input"
        tabIndex="1" />
    </div>
  </div>
  <div
    className="submit-button"
    onClick={[Function]}>
     Submit! 
  </div>
</div>
`;

exports[`test number of boxes equal to synonyms 2`] = `
<div
  className="input-container">
  <div
    className="input-flex-container">
    <div
      className="input-box"
      style={
        Object {
          "background": "white",
        }
      }>
      <input
        className="the-input"
        tabIndex="1" />
    </div>
    <div
      className="input-box"
      style={
        Object {
          "background": "white",
        }
      }>
      <input
        className="the-input"
        tabIndex="1" />
    </div>
    <div
      className="input-box"
      style={
        Object {
          "background": "white",
        }
      }>
      <input
        className="the-input"
        tabIndex="1" />
    </div>
    <div
      className="input-box"
      style={
        Object {
          "background": "white",
        }
      }>
      <input
        className="the-input"
        tabIndex="1" />
    </div>
  </div>
  <div
    className="submit-button"
    onClick={[Function]}>
     Submit! 
  </div>
</div>
`;
